pkg/validation/internal/multiarch.go

```diff
@@ -0,0 +1,513 @@
+const operatorFrameworkArchLabel = "operatorframework.io/arch."
+
+// operatorFrameworkSoLabel defines the label used to the store the supported So on CSV
+const operatorFrameworkSoLabel = "operatorframework.io/so."
+
+// MultiArchValidator store the data to perform the tests
+type MultiArchValidator struct {
+	// InfraArchLabels store the arch labels (i.e amd64, ppc64le) from
+	// operatorframework.io/arch.<GOARCH>: supported
+	InfraArchLabels []string
+	// InfraSOLabel store the OS labels from
+	// operatorframework.io/so.<GOARCH>: supported
+	InfraSOLabels []string
+	// Images stores the images defined in the bundle with the platform.arch supported
+	Images map[string][]Platform
+	// AllArchFound contains a map of the arch types found
+	AllArchFound map[string]string
+	// AllOsFound contains a map of the so(s) found
+	AllOsFound map[string]string
+	// Store the bundle load
+	Bundle *manifests.Bundle
+	// ContainerTool defines the container tool which will be used to inspect the images
+	ContainerTool string
+	// Warns stores the errors faced by the validator to return the warnings
+	Warns []error
+}
+
+// ManifestInspect store the data obtained by running container-tool manifest inspect <IMAGE>
+type ManifestInspect struct {
+	ManifestData []ManifestData `json:"manifests"`
+}
+
+// ManifestData store the platforms
+type ManifestData struct {
+	Platform Platform `json:"platform"`
+}
+
+// Platform store the Architecture and OS supported by the image
+type Platform struct {
+	Architecture string `json:"architecture"`
+	OS           string `json:"os"`
+}
+
+func multipleArchitecturesValidate(objs ...interface{}) (results []errors.ManifestResult) {
+	// Obtain the k8s version if informed via the objects an optional
+	var containerTool = ""
+	for _, obj := range objs {
+		switch obj.(type) {
+		case map[string]string:
+			containerTool = obj.(map[string]string)[ContainerToolsKey]
+			if len(containerTool) > 0 {
+				break
+			}
+		}
+	}
+
+	for _, obj := range objs {
+		switch v := obj.(type) {
+		case *manifests.Bundle:
+			results = append(results, validateMultiArchWith(v, containerTool))
+		}
+	}
+	return results
+}
+
+func validateMultiArchWith(bundle *manifests.Bundle, containerTool string) errors.ManifestResult {
+	result := errors.ManifestResult{}
+	if bundle == nil {
+		result.Add(errors.ErrInvalidBundle("Bundle is nil", nil))
+		return result
+	}
+
+	result.Name = bundle.Name
+
+	if bundle.CSV == nil {
+		result.Add(errors.ErrInvalidBundle("Bundle csv is nil", bundle.Name))
+		return result
+	}
+
+	containerTool, err := validateContainerTool(containerTool)
+	if err != nil {
+		result.Add(errors.ErrFailedValidation(err.Error(), bundle.CSV.GetName()))
+		return result
+	}
+
+	multiArchValidator := MultiArchValidator{Bundle: bundle, ContainerTool: containerTool}
+	multiArchValidator.validate()
+
+	for _, err := range multiArchValidator.Warns {
+		result.Add(errors.WarnFailedValidation(err.Error(), bundle.CSV.GetName()))
+	}
```

Comment:
I think this section is a little bit confusing.

1. Why is `MultiArchValidator` exported when its primary function, `validate` is not?
2. Why are `Warns` read off the struct instead of being returned?

Could we do something like this?

```go
for _, err := range ValidateMultiArchBundle(bundle, containerTool) {
    // add the warn to the result
}

func ValidateMultiArchBundle(bundle *manifest.Bundle, containerTool string) []error {
    data := &multiArchValicatorData{
        AllArchFound: make(map[string]string),
        // etc
    }
    data.getInfraLabelsFromCSV(bundle)
    return data.checks() // this will return the warnings instead of mutating the struct
}
```

This gives us:

1. Single entrypoint to validate function, no instantiating struct then calling validate.

2. More self documenting since consumers of the API won’t need to know to read struct fields to get results. 

3. Don’t need to use single use struct.

Reply:
done some improvements based on your feedback.